### PR TITLE
[v0.3 DENAS] refine e2eaiok common lib

### PR DESCRIPTION
**Pull request by Peach-He**
common lib
* common
  * trainer
    * data_builder.py
      * prepare_dataset
      * get_dataloader
    * model_builder.py
      * _init_model
      * load_model
      * create_model
    * torch_trainer.py
      * fit
      * train_one_epoch
      * evaluate
    * data
      * data_builder_cv/data_builder_nlp/data_builder_asr ...
    * model
      * model_builder_cv/nlp/asr ...
* DeNas
  * asr/cv/nlp
    * cv/nlp/asr_trainer.py -- inherit from common/trainer/torch_trainer.py
  * train.py -- call common functions to build training pipeline

high level architecture of AIOK trainer common lib @xuechendi

### CODE REVIEW

File: e2eAIOK/common/trainer/train.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import torch
```
Comment (tianyil1) [inline e2eAIOK/common/trainer/train.py:4]: Shall we put the trainer in the common folder, or define the separate trainer for DENAS and Model Adaptor?

**Review by tianyil1 — COMMENTED** [review]
Below is my comment, please check it. Thx.

File: e2eAIOK/common/trainer/train.py (hunk 1 above)
Comment (Peach-He) [reply e2eAIOK/common/trainer/train.py:4]: the trainer in common folder is the base trainer class, for training scripts, DENAS and Mode Adaptor can define their own training scripts. Maybe it is better to put train.py in DENAS/Model Adaptor folder

File: e2eAIOK/common/trainer/model_builder.py [hunk 2]
```diff
@@ -15,38 +15,40 @@ def __init__(self, cfg):
...
+        """
+        raise NotImplementedError("_init_model is abstract.")
+    
+    def _post_process(self):
+        """
+            post work after create model
+        """
+        print(f"model created: {self.model}")
+
     def create_model(self):
```
Comment (XinyaoWa) [inline e2eAIOK/common/trainer/model_builder.py:34]: 1. Can we merge create_model() and __init__() together, in that way we can create model when initialize the Builder.
2. We can add a parameter "model" to __init__() to load extra models.

File: e2eAIOK/common/trainer/utils.py [hunk 3]
```diff
@@ -173,29 +183,25 @@ def log_every(self, iterable, print_freq, header=None):
...
-    all_operations = {'criterion':criterion,'optimizer':optimizer,'lr_scheduler':lr_scheduler}
-    return all_operations
+    return optimizer
 
 def create_metric(output, target, cfg):
     if cfg.eval_metric == "accuracy":
         acc1, acc5 = accuracy(output, target, topk=(1, 5))
         metric = {'acc1':acc1,'acc5':acc5}
-        return metric
-
```
Comment (XinyaoWa) [inline e2eAIOK/common/trainer/utils.py:201]: 1. These functions are duplicated in torch_trainer.py, maybe we can only keep one
2. Suggest move utils.py to util file

File: e2eAIOK/common/trainer/train.py (hunk 1 above)
Comment (tianyil1) [reply e2eAIOK/common/trainer/train.py:4]: Yes, I think putting the train.py in DENAS/Model Adaptor folder sounds good.

File: e2eAIOK/common/trainer/model_builder.py (hunk 2 above)
Comment (tianyil1) [reply e2eAIOK/common/trainer/model_builder.py:34]: For the 2nd suggestion, maybe we can do it like "__init__(self, *kargs)" for extra arguments. Thx.

Comment (XinyaoWa) [reply e2eAIOK/common/trainer/model_builder.py:34]: But we need to use it in the next step, such as:
> if model is not None:
>       self.model = model
> else:
>      self.model = _init_model()

I this case, I suggest model can work as a normal parameter.

File: e2eAIOK/common/trainer/train.py (hunk 1 above)
Comment (Peach-He) [reply e2eAIOK/common/trainer/train.py]: move train.py to DENAS/Model Adaptor folder

File: e2eAIOK/common/trainer/model_builder.py (hunk 2 above)
Comment (Peach-He) [reply e2eAIOK/common/trainer/model_builder.py:35]: add "model" parameter in init() to load pre-trained model

File: e2eAIOK/common/trainer/utils.py (hunk 3 above)
Comment (Peach-He) [reply e2eAIOK/common/trainer/utils.py:201]: move functions from torch_trainer to utils.py and move utils.py to utils folder

**Review by XinyaoWa — APPROVED** [review]

**Review by xuechendi — APPROVED** [review]
LGTM

**Pull request merged**